[WIP]Refresh framework for MaterializedView

Astralidea opened the pull request:
## What type of PR is this：
- [ ] bug
- [x] feature
- [ ] enhancement
- [ ] others

## Which issues of this PR fixes ：
<!--
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #4796

## Background
The old materialization attempts to only support a single table and do it in a synchronous refresh manner by mapping the partitions of the single table into a one-to-one mapping of the Base table and the MV table.

When the old materialized view is building a table, the process of creating the materialized view is asynchronous, but the asynchronous scheduling framework relies on the MasterDaemon to be inefficient.

In order to cope with the requirements of asynchronous refresh and asynchronous triggering of multi-table materialized views. Therefore, it is necessary to build a new refresh the framework.

## Scenario
This framework currently only focuses on asynchronous refreshes.
1.For real-time scenarios, users generally build multiple tables, and then periodically execute an SQL to refresh the target table.
2.For offline scenarios, the user will periodically refresh all of them once a day.

## Design
We mainly divide refresh scenarios into two categories, one is periodic refresh, and the other is due to table creation, Base table change (data update) and user-triggered refresh.

All our refreshes are done with MaterializedViewJobManager.
For periodic refresh, it must be automatically triggered by the system at present. We provide the registerScheduledJob interface to register a periodic refresh task, and the deregisterScheduledJob interface to deregister a periodic task. For other types of triggered refreshes, we provide addPendingJob and cancelJob interfaces to support, and periodic tasks are essentially called addPendingJob to add tasks for management.

<img width="665" alt="image" src="https://user-images.githubusercontent.com/4392280/164009306-12e520ba-158c-4bf4-8f57-a2b92efcfa45.png">

Finally, it will be added to the pendingJob queue uniformly, and a dispatcher thread will perform a scheduling operation every 1s. This scheduling currently uses the minimum heap that comes with JDK, which can be replaced with a time wheel later, and can only be generated according to each MV. The logic of a running job is executed in the unrestricted refreshTaskPool, and a concurrency limit will be added to this operation later.

The same tasks that are automatically generated will also be merged here. For tasks manually generated by the user, we will not merge them.

It is also necessary to consider the logic of resuming the execution of some tasks after the FE restarts, and the logic of task cleaning. At present, this part of the logic has not been written, and will be added later. Comments are used in the code to mark where the log needs to be written.

Astralidea committed:
Refresh framework for MaterializedView

Astralidea committed:
Refresh framework for MaterializedView

Seaven committed:
Update mysql/es external table output row count (#5152)

wanpengfei-git commented:
#### [FE PR Coverage check]

:disappointed: **fail** : 241 / 336 (71.73%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/mv/MaterializedViewRefreshTask.java|0|18|00.00%|
|:large_blue_circle:|com/starrocks/mv/MaterializedViewPartitionRefreshTask.java|0|28|00.00%|
|:large_blue_circle:|com/starrocks/catalog/Catalog.java|0|1|00.00%|
|:large_blue_circle:|com/starrocks/mv/MaterializedViewSchedulerInfo.java|10|20|50.00%|
|:large_blue_circle:|com/starrocks/mv/MaterializedViewRefreshJob.java|56|71|78.87%|
|:large_blue_circle:|com/starrocks/mv/MaterializedViewJobManager.java|19|24|79.17%|
|:large_blue_circle:|com/starrocks/statistic/Constants.java|19|22|86.36%|
|:large_blue_circle:|com/starrocks/mv/MaterializedViewJobScheduler.java|108|122|88.52%|
|:large_blue_circle:|com/starrocks/mv/MaterializedViewRefreshJobBuilder.java|29|30|96.67%|

ZiheLiu committed:
Fix: use strict mode of DateTimeFormatter for parsing string to date (#5146)

murphyatwork committed:
fix #4996 exchange operator exception handling (#5157)

* fix #4996

* handle exception in local exchange sink operator

Youngwb committed:
Fix group expression not replace lowestCostExpressions in the group when group merge (#5007)

kangkaisen committed:
Fix unknow error when CallOperator as predicate and the table has materialized view (#5148)

stdpain committed:
return error when string function generate a large column (#5159)

trueeyu committed:
Add TCMALLOC_HEAP_LIMIT_MB set check before be start (#5171)

sevev committed:
Add support for the `enable_persistent_index` modification (#4958)

We add `enable_persistent_index` in properties to decide whether to use the persistent index in the PrimaryKey tablet. `enable_persistent_index` is specified when creating a table, and this pr adds support for the enable_persistent_index modification. We can use Alter table xxx set ("enable_persistent_index" = "false(true)") to modify it.

Youngwb committed:
Fix double cast to string use scientific notation in Fe (#5181)

xlwh committed:
Fix some typo and cleanup unused include (#5184)

murphyatwork committed:
fix #5028: return null if AssertNumRows input empty (#5166)

xlwh committed:
Remove directory be/test/storage/vectorized (#5136)

xiaoyong-z committed:
[Enhancement] optimize _get_segment_iterator_num in compaction (#5063)

* [Enhancement] optimize _get_segment_iterator_num in compaction
Compaction needs to choose the compaction algorithm(vertical of horizontal)
based on the `segment_iterator_num`.
In previous implementation, we need to first
construct a tablet reader, then construct all segment iterators,
the segment_iterator_num equals to the number of segment iterators.
The cost of computing segment_iterator_num is very high.
In fact, we can directly get the segment_iterator_num from the meta of rowsets.
This method will be much more efficient than the original one.

murphyatwork committed:
[Opt merge part1] Introduce columnwise merge algorithm (#5066)

Implement a column-wise merge algorithm:

* Merge the first column, record the equal rows into a equal-range
* Merge the second column within the equal-range of previous column
* Repeat until no equal-range or last column

The advantage of this merge algorithm is:

* Eliminate the compare function call during merge
* Improve the memory access efficiency by sequential iterate two columns

Linkerist committed:
remove the restriction of modify column clause to support modify column for primary key model. (#5199)


Primary key table do not support modify column for now.
```
mysql> create table orders (
    ->     dt date NOT NULL,
    ->     order_id bigint NOT NULL,
    ->     user_id int NOT NULL,
    ->     merchant_id int NOT NULL,
    ->     good_id int NOT NULL,
    ->     good_name string NOT NULL,
    ->     price int NOT NULL,
    ->     cnt int NOT NULL,
    ->     revenue int NOT NULL,
    ->     state tinyint NOT NULL
    -> ) PRIMARY KEY (dt, order_id)
    -> DISTRIBUTED BY HASH(order_id) BUCKETS 4
    -> PROPERTIES("replication_num" = "1");
Query OK, 0 rows affected (0.12 sec)

mysql> ALTER table orders MODIFY COLUMN price BIGINT;
ERROR 1064 (HY000): Primary key table do not support modify column
```

just remove the restriction of modify column clause, in fact, the implement was supported already.

decster committed:
Add tablet_max_pending_versions config; improve and bugfix some logs (#5168)

liuyehcf committed:
Add join on predicate property info (#4927)

Seaven committed:
CTE produce output optimize (#5093)

Astralidea committed:
add meta info

Astralidea committed:
add meta info

stdpain committed:
Fix const value handle in UDAF (#5126)

now our Aggregator only unpack the first parameter. so if we create a UDAF parameter greater than 2, we will got a exception

mv cast_to_jvalue/release_jvalue to java_data_converter.cpp
handle const parameter in Java UDAF

miomiocat committed:
Fix wrong min-max filter result of parquet reader (#4930)

miomiocat committed:
Fix missing materialized column for hudi table (#5119)

wanpengfei-git commented:
#### [FE PR Coverage check]

:disappointed: **fail** : 261 / 455 (57.36%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/mv/MaterializedViewRefreshTask.java|0|18|00.00%|
|:large_blue_circle:|com/starrocks/persist/EditLog.java|0|22|00.00%|
|:large_blue_circle:|com/starrocks/mv/MaterializedViewPartitionRefreshTask.java|0|28|00.00%|
|:large_blue_circle:|com/starrocks/journal/JournalEntity.java|0|9|00.00%|
|:large_blue_circle:|com/starrocks/catalog/Catalog.java|0|2|00.00%|
|:large_blue_circle:|com/starrocks/mv/MaterializedViewRefreshJobStatusChange.java|6|23|26.09%|
|:large_blue_circle:|com/starrocks/mv/MaterializedViewSchedulerInfo.java|11|26|42.31%|
|:large_blue_circle:|com/starrocks/mv/MaterializedViewJobScheduler.java|121|175|69.14%|
|:large_blue_circle:|com/starrocks/mv/MaterializedViewRefreshJob.java|56|76|73.68%|
|:large_blue_circle:|com/starrocks/mv/MaterializedViewJobManager.java|19|24|79.17%|
|:large_blue_circle:|com/starrocks/statistic/Constants.java|19|22|86.36%|
|:large_blue_circle:|com/starrocks/mv/MaterializedViewRefreshJobBuilder.java|29|30|96.67%|

rickif committed:
BugFix: consumer not been scheduled when consumers of group larger than three (#5132)

The BE uses a thread pool to execute the consumers of the consumer group, while the size of the thread pool is three. When the number of consumers is larger than 3, some consumers have to wait for the availability of the thread pool. The delay of execution may lead to the task running time beyond routine_load_task_consume_second, and then the hanging consumer would be canceled before being executed.

rickif committed:
[Enhancement] reuse json parser to speed up json parsing (#5010)

This PR reuses the JSON parser and buffer to get better performance, which is recommended in
https://github.com/simdjson/simdjson/blob/master/doc/performance.md

## Benchmark
Type: Routine Load
Source: message in JSON format; 300 bytes per message; 1 partition in Kafka
StarRocks Config: 3 BE with 3 replica

### Throughput
| version | throughput|
|--| --|
|perf/reuse-parser|43.86 MB/s|
| main| 31.33 MB/s|

note: The throughput is computed by received_bytes / consume_time(ms)
```
I0412 19:36:57.615734 11959 data_consumer_group.cpp:131] consumer group done: f84a5ccc0736318a-cf93f37dcd661b9b. consume time(ms)=3011, received rows=786397, received bytes=108522786, eos: 0, left_time: -11, left_bytes: 415765214, blocking get time(us): 1293085, blocking put time
(us): 940130
```

miomiocat committed:
Fix a minor typo of thirdparty script (#5226)

imay committed:
Remove unused boost thread includes (#5230)

gengjun-git committed:
BugFix: Colocate table's tablet cannot be repaired after be is dropped (#5129)

We should not repair the replica that is not in the colocate backends set, this kind of replica should be dropped.

meegoo committed:
Fix asan check crash since selected column order not match (#5147)

meegoo committed:
Avoid unnecessary error log generation overhead (#5238)

stdpain committed:
[optimize] optimize cast string to integer (#5228)

1FE 4BE 1PARALLEL: SSB_50g

baseline: 0.80 sec, optimize: 0.41 sec

stdpain committed:
[Mem] add exception check when evaluate conjuncts (#5254)

add exception check when evaluate conjuncts

HangyuanLiu committed:
Fix subquery correlated can resolved multi level outer layer #5224

wyb committed:
Backup restore compatible with apache doris (#5227)

Backup table from apache doris 0.14, and restore to StarRocks will throw error.
Because tabletsOrder is stored as long in apache doris and is parsed in StarRocks as String.

In addition, |tablets| is a map in backup __info_<timestamp>.<md5> file, so tablet order is not guaranteed.

So we use Long in tabletsOrder for compatibility with apache doris,
and serialize tabletsOrder to guarante the correct tablet order.

Astralidea committed:
support retryJob

wanpengfei-git commented:
#### [FE PR Coverage check]

:disappointed: **fail** : 309 / 492 (62.80%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/persist/EditLog.java|0|22|00.00%|
|:large_blue_circle:|com/starrocks/mv/MaterializedViewPartitionRefreshTask.java|0|19|00.00%|
|:large_blue_circle:|com/starrocks/journal/JournalEntity.java|0|9|00.00%|
|:large_blue_circle:|com/starrocks/catalog/Catalog.java|0|2|00.00%|
|:large_blue_circle:|com/starrocks/mv/MaterializedViewRefreshJobStatusChange.java|6|23|26.09%|
|:large_blue_circle:|com/starrocks/mv/MaterializedViewRefreshTask.java|8|20|40.00%|
|:large_blue_circle:|com/starrocks/mv/MaterializedViewSchedulerInfo.java|11|26|42.31%|
|:large_blue_circle:|com/starrocks/mv/MaterializedViewJobScheduler.java|137|195|70.26%|
|:large_blue_circle:|com/starrocks/mv/MaterializedViewRefreshJob.java|72|91|79.12%|
|:large_blue_circle:|com/starrocks/mv/MaterializedViewJobManager.java|21|26|80.77%|
|:large_blue_circle:|com/starrocks/mv/MaterializedViewTaskException.java|5|6|83.33%|
|:large_blue_circle:|com/starrocks/statistic/Constants.java|20|23|86.96%|
|:large_blue_circle:|com/starrocks/mv/MaterializedViewRefreshJobBuilder.java|29|30|96.67%|

Astralidea committed:
Fix bug type convert and remove findCommonType (#5257)

the cause of this bug is because k6-1 expr is both tinyint type.
and use findCommonType will return INVALID type.
but in new optimizer use getCommonType to issue this problem so we remove old findCommonType code and use new one.

xlwh committed:
Move runtime/vectorized to runtime/ (#5183)

ylcq committed:
[SR GeekCamp] Improve schema scan performance #4028 (#5124)

meegoo committed:
Fix broker load failed when use NEGATIVE and value is tinyint/smallint/int (#5271)

murphyatwork committed:
fix #5250: append_by_permutation bug for BinaryColumn (#5267)

silverbullet233 committed:
Skip filtering unused column if outputColumns of ScanOperator is empty (#5206)

kangkaisen committed:
Fix BE crash when enable_filter_unused_columns_in_scan_stage is ture for double column (#5212)

murphyatwork committed:
fix compile error introduced by vectorized namespace (#5273)

Linkerist committed:
Do rocksdb compact before data dir load. (#5076)

Astralidea committed:
add UT & refactor

Astralidea committed:
add UT & refactor

decster committed:
make update mem limit configurable (#5258)

This PR makes update memory limit configurable, adds a new BE config update_memory_limit_percent.

Astralidea commented:
run starrocks_fe_unittest

Astralidea commented:
run starrocks_fe_unittest

wanpengfei-git commented:
#### [FE PR Coverage check]

:disappointed: **fail** : 312 / 486 (64.20%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/persist/EditLog.java|0|22|00.00%|
|:large_blue_circle:|com/starrocks/mv/MaterializedViewPartitionRefreshTask.java|0|19|00.00%|
|:large_blue_circle:|com/starrocks/journal/JournalEntity.java|0|9|00.00%|
|:large_blue_circle:|com/starrocks/catalog/Catalog.java|0|1|00.00%|
|:large_blue_circle:|com/starrocks/mv/MaterializedViewRefreshJobStatusChange.java|6|23|26.09%|
|:large_blue_circle:|com/starrocks/mv/MaterializedViewRefreshTask.java|8|20|40.00%|
|:large_blue_circle:|com/starrocks/mv/MaterializedViewSchedulerInfo.java|11|20|55.00%|
|:large_blue_circle:|com/starrocks/mv/MaterializedViewJobManager.java|138|196|70.41%|
|:large_blue_circle:|com/starrocks/mv/MaterializedViewJobExecutor.java|18|23|78.26%|
|:large_blue_circle:|com/starrocks/mv/MaterializedViewRefreshJob.java|77|94|81.91%|
|:large_blue_circle:|com/starrocks/mv/MaterializedViewTaskException.java|5|6|83.33%|
|:large_blue_circle:|com/starrocks/statistic/Constants.java|20|23|86.96%|
|:large_blue_circle:|com/starrocks/mv/MaterializedViewRefreshJobBuilder.java|29|30|96.67%|

mergify[bot] committed:
Add some info log about metastore event processor (#5120) (#5140)

stdpain committed:
support sort in low cardinality optimize (#4989)

stdpain committed:
Fix incorrect result set under concurrent query (#5281)

imay committed:
Remove AnyVal (#5278)

trueeyu committed:
Add one check for ArrayColumn (#5266)

* Add one check for ArrayColumn

* add comment

silverbullet233 committed:
fix data column size error when creating const column in decimal_round function (#5086)

Youngwb committed:
Support query dump for query view stmt and insert stmt (#5158)

imay committed:
Remove os_info (#5291)

ABingHuang committed:
Fix fail to prepare tablet reader bug (#5155)

When publish stuck in one replica and the last failed version is > 0,
the replica will not be treated as an error replica, and FE will update
the replica's version to the newest visible version of partition when the replica
report. and the query will fail because missing version because the publish
the version does not finish in BE; the version is not queryable.
And previously, the quorum success of the published version depended on the tablet report.
I added an optimization to decide quorum success based on the publishing version task.
It will boost the publish version process from 10s+ to 0.03s.

chaoyli committed:
[Enhancement] Remove unused compaction log (#5248)

HangyuanLiu committed:
When a subquery is included in an aggregate function, the subquery is parsed multiple times (#5095)

dirtysalt committed:
By default reserve 1MB bytes for parquet column reader (#5294)

dirtysalt committed:
By default disable fetching ec2-metadata (#5295)

wyb committed:
New backup restore compatible with apache doris (#5297)

but old version StarRocks still use string to deserialize tablets_order,
this will cause ArrayStoreException when rolling back to old version.

So we use a new solution for compatibility.
1. Keep using string to serialize tablets_order as before.
2. Deserialize string or long tablets_order by uniform way.

gengjun-git committed:
BugFix: Storage medium cannot be changed to hdd when cool_down_time is reached #5232

Change storage_medium to HDD instead of default_storage_medium when cool_down_time is reached.

ZiheLiu committed:
Opt: rewrite zonemap predicates only once (#5286)

liuyehcf committed:
Fix SinkBuffer block problem (#5144)

decster committed:
Refactor and add more logging in remove_expired_versions and get_applied_rowsets (#5272)

This PR refactor and add more logging in remove_expired_versions and get_applied_rowsets

dirtysalt committed:
orc List/Map Vector Batch support filter (#5318)

sevev committed:
BugFix: fix StarRocks#5039: Be upgrade failed (#5314)

If we create a cluster with apache-doris-0.13, we can upgrade it to starrocks-1.19.x. But when we upgrade it from starrocks-1.19.x to starrocks-2.0.x, be may be failed.

The reason is we add `id` field to `TabletSchemPB` in starrocks-2.0 to support Shared-Schema which conflicts with apache-doris's `TabletSchemaPB`

`TabletSchemaPB` of starrocks-2.0
```
message TabletSchemaPB {
    optional KeysType keys_type = 1;    // OLAPHeaderMessage.keys_type
    repeated ColumnPB column = 2;   // OLAPHeaderMessage.column
    optional int32 num_short_key_columns = 3;   // OLAPHeaderMessage.num_short_key_fields
    optional int32 num_rows_per_row_block = 4;  // OLAPHeaderMessage.num_rows_per_data_block
    optional CompressKind compress_kind = 5; // OLAPHeaderMessage.compress_kind
    optional double bf_fpp = 6; // OLAPHeaderMessage.bf_fpp
    optional uint32 next_column_unique_id = 7; // OLAPHeaderMessage.next_column_unique_id
    optional bool DEPRECATED_is_in_memory = 8 [default=false];
    optional int64 id = 9;
}
```

`TabletSchemaPB` of apache-doris-0.13
```
message TabletSchemaPB {
    optional KeysType keys_type = 1;    // OLAPHeaderMessage.keys_type
    repeated ColumnPB column = 2;   // OLAPHeaderMessage.column
    optional int32 num_short_key_columns = 3;   // OLAPHeaderMessage.num_short_key_fields
    optional int32 num_rows_per_row_block = 4;  // OLAPHeaderMessage.num_rows_per_data_block
    optional CompressKind compress_kind = 5; // OLAPHeaderMessage.compress_kind
    optional double bf_fpp = 6; // OLAPHeaderMessage.bf_fpp
    optional uint32 next_column_unique_id = 7; // OLAPHeaderMessage.next_column_unique_id
    optional bool is_in_memory = 8 [default=false];
    optional int32 delete_sign_idx = 9 [default = -1];

}
```

When we upgrade to starrocks-2.0.x, `delete_sign_idx` will be considered as `id` and we can't guarantee that it's value is meaningful.

Youngwb committed:
Fix LogicValueOperator misjudgment of equality when columnRefSet and rows are empty (#5324)

murphyatwork committed:
Fix sort expression bug (#5321)

imay committed:
Build minizip in thirdparty (#5315)

Currently, StarRocks copy minizip codes in self reop. However it is not
necessary, because minizip is included in libz. So we only need to build
it in libz, then we can use it as third-party depedency.
This is the first PR to add minizip in third-part build script. After
this PR, I will remove code in be/src/util/minizip and make StarRocks
depend minizip in libz.

xlwh committed:
Fix thirdparty issue for pentelemetry (#5332)

sduzh committed:
Move lru_cache.{h,cpp} to be/src/util (#5331)

murphyatwork committed:
[Opt merge part2] introduce cascade merge algorithm (#5067)

Introduce a cascade merge algorithm:

* Two-way stream merge: cutoff streaming chunk into small morsels, merge each morsel individually
* Multi-way cascade merge: implement cascade merge based on two-way merge

Youngwb committed:
Fix group merge phase not update the GroupExpression state which will merge other GroupExpression (#5211)

ZiheLiu committed:
Fix: sum to double not int64 for avg(int) without group by (#5325)

xiaoyong-z committed:
[Enhancement] Early release stale rowsets after compaction.  (#5056)

* [optimize] Early release stale rowsets after compaction.
After compaction, the stale rowsets will still stay in memory for a long time before gc thread free them. And these stale rowsets take up a lot of memory. In fact, there may be a few old readers still read the stale rowsets, all new readers will read the new rowset generated by the compaction, so we can free the stale rowsets if there are no readers on it. Experiments show that this optimization significantly reduces the memory footprint of rowsets meta data.

imay committed:
Remove unused PlanNode (#5344)

Seaven committed:
Fix decimal expression can't reuse (#5221)

dirtysalt committed:
ORC List/Map Vector Batch support filter (#5337)

murphyatwork committed:
fix #5259: catch exception in boost::tokenizer (#5345)

xlwh committed:
Fix issue for opentelemetry lib compile (#5356)

miomiocat committed:
Support describe all and show proc command for hudi table (#5242)

dirtysalt committed:
Support _col0,_col1 for orc late materialization (#5355)

rickif committed:
[BugFix] json stream load failed with Transfer-Encoding: chunked set (#5330)

When `Transfer-Encoding: chunked` is set, the content of the HTTP body cannot be known from the HTTP header `Content-Length`.
However, the previous implement of `StreamLoadPipe` need the http body size for initializing `_total_length`, and the `StreamLoadPipe:read_one_message` also needs `_total_length`.

This PR includes the following modifications:
1. remove the `StreamLoadPipe`'s dependency on HTTP body size as the initial size
2. redefine the semantics of `StreamLoadPipe:read_one_message` as `returns data that is written by append in one time`
3. new unit tests for `StreamLoadPipe:read_one_message`

The `StreamLoadPipe` is a pipe to store data from stream load. Every time we use `StreamLoadPipe::append` for JSON, we must append a complete JSON payload. So, it's OK to parse the data as JSON returned by  `StreamLoadPipe:read_one_message`.

stdpain committed:
rethrow InvocationTargetException in Java UDAF (#5305)

ref:
https://stackoverflow.com/questions/6388051/java-invocationtargetexception

dreamay committed:
[SR GeekCamp]New Parser and Analyzer support ADMIN SET related statement (#5134)

ylcq committed:
[SR GeekCamp] Improve schema scan performance #4028 (#5320)

aaawuanjun committed:
[SR GeekCamp] Support bitmap as a primitive type【Easy】 (#5078)

Thie PR adds support for bitmap type to PRIMARY/DUPLICATE/UNIQUE key table

goodqiang committed:
Fix BE crashes #5346 (#5368)

stdpain committed:
optimize uuid() implements (#5109)

HangyuanLiu committed:
Support Insert table with partition in new parser #5377

Astralidea committed:
[WIP]upgrade jprotobuf-rpc to 4.2.1 remove dependency of spring (#5032)

For jprotobuf-rpc-core 4.2.1 it remove spring dependency for jprotobuf-rpc-core-test & junit already excluded.

sduzh committed:
Remove useless const keyword (#5380)

miomiocat committed:
Fix refresh cache issue for hudi table (#5382)

Seaven committed:
Fix equals columns check bug (#5352)

stdpain committed:
fix rolling upgrade problems (#5393)

because old lowcardinality optimization rewrite origin_expr return type
to int type, which will cause DictExpr return a string column

sduzh committed:
Remove memory copy in RowsetMetaManager::traverse_rowset_meta (#5397)

imay committed:
Put some function implementation to column.cpp (#5401)

sduzh committed:
Remove useless argument of RowsetMetaManager::save (#5405)

wyb committed:
Bugfix: check array state column nullable when ReplaceAggregator appending data (#5414)

When loading array replace if not null column and the same key has only null data,
the array state column will be null, and it will cause be crashes if not check nullable.

sduzh committed:
Reduce memory copy when deserializing tablet meta (#5418)

xlwh committed:
fix the static check bug of tscancode (#5426)

zombee0 committed:
Delete virtual on append_selective (#5429)

Signed-off-by: zombee0 <flylucas_10@163.com>

xlwh committed:
Fix scan cpp check error (#5428)

xlwh committed:
split be/src/util/tdigest.h into headers and cpp files (#5424)

mxdzs0612 committed:
Support custom catalog for iceberg external table (#5117)

Add support for custom catalog which can be defined by users themselves when creating iceberg external table.

The custom catalog should be in the form of IcebergHiveCatalog, in other words extending BaseMetastoreCatalog and implementing IcebergCatalog. The catalog JAR should be placed into each fe/lib directory, and FE has to be restarted before custom catalog works.

Usage:
```sql
CREATE EXTERNAL RESOURCE "iceberg0"
PROPERTIES (
  "type" = "iceberg",
  "starrocks.catalog-type"="CUSTOM",
  "iceberg.catalog-impl"="{The full class name of custom catalog}"
);
```
Extra config users defined can be added in table properties when executing CREATE EXTERNAL TABLE, see #2225.

caneGuy committed:
Move data lake pushdown rule into mergepredicate rule (#5391)

wangruin committed:
Optimize Build user when displaying fe or be version (#5437)

zddr committed:
[SR GeekCamp]merge join fe （issue:4036） (#4882)

Fe support SortMergeJoin

zombee0 committed:
pass std::string_view by value (#5431)

murphyatwork committed:
[Opt merge part4] refactor SortContext (#5289)

Refactor the merge sort in SortContext, to take the advantage of column-wise merge.

dirtysalt committed:
Refactor parquet reader (file layout not functionality) (#5443)

ABingHuang commented:
there are some points that need to be consider:
1. should connect with create materialized view
2. should consider load image and read a Materialized view
3. should consider the concurrency control

liuyehcf committed:
Add ut for decimal truncate/round (#5253)

dirtysalt committed:
Refactor orc chunk reader (#5447)

liuyehcf committed:
Fix nullable attribute of SlotDescriptor (#5298)

sduzh committed:
Remove unused class FileReader (#5454)

dirtysalt committed:
Move exec/parquet to formats/parquet (#5452)

rickif committed:
[Enhancement] construct row in json object order when column is sparse (#5357)

The previous implement constructed a row from JSON in column order. The JSON scanner iterates the column descriptor and looks up the column name in the JSON object. When the number of columns is much more than the JSON field, there're many useless looking up of the JSON object. The new implementation uses a smarter strategy to construct the columns: If the number of columns is much more (2x) than the number of object fields, lookup the object field key in the dictionary of the column. Otherwise, look up the column name in the JSON object.

## Benchmark
load type: Stream Load
source file: 65MB in total, json array format, 6KB / object, 64 fields /object
table: 330 columns, 10 buckets
| version | throughput |
| -- | -- |
| main | 2.5MB/s|
| c8d00b2e4@perf/sparse-column| 18.2MB/s |

padmejin committed:
[Enhancement] Check user identity for `show grants` & support string literal role name for `create role` and `drop role` (#5390)

1. So far using `show grants` for a non-existent user will end up showing a table with all cells 'null'. Fix it by raising an exception in this scenario.
2. Allow usage of `create role 'role_name'` by changing role CUP type from ident to ident_or_text.

murphyatwork committed:
Revert "[Opt merge part4] refactor SortContext (#5289)" (#5467)

This reverts commit 92c4ba6bd580e2366065ee18a138676155a326b8.

sduzh committed:
Rename rowset_id_v2 to rowset_id (#5413)

kangkaisen committed:
Pass const args to merge phase aggregator in BE (#5381)

blackstar-baba committed:
[SR GeekCamp] Reorder predicates based on selectivity (#4869)

Youngwb committed:
Fix show table from db using back quote (#5469)

stdpain committed:
support fixed hashmap in Aggregator (#5444)

dirtysalt committed:
Add more tests into starrocks_test (#5455)

sduzh committed:
Refine StreamLoadPipe::append_and_flush() (#5449)

dirtysalt committed:
fix BE ut cases (#5481)

kangkaisen committed:
Fix compile (#5482)

choury committed:
Add window_funnel aggregate function (#4985)

Co-authored-by: honglichang <honglichang@tencent.com>

hiliuxg committed:
[SR GeekCamp] Create table statement support for List partition  (#5284)

1. Create Table for single column partition

CREATE TABLE t_recharge_detail_single(
    id bigint  ,
    user_id  bigint  ,
    recharge_money decimal(32,2) , 
    province varchar ,
    dt varchar
) ENGINE=OLAP
DUPLICATE KEY(id)
PARTITION BY LIST (province) (
   PARTITION p1 VALUES IN ("beijing","chongqing"),
   PARTITION p2 VALUES IN ("shanghai","tianjing")
)
DISTRIBUTED BY HASH(id) BUCKETS 1
PROPERTIES(
    "replication_num" = "1"
)

2. Create Table for multi columns partition

CREATE TABLE t_recharge_detail_multi(
    id bigint  ,
    user_id  bigint  ,
    recharge_money decimal(32,2) , 
    province varchar ,
    dt varchar
) ENGINE=OLAP
DUPLICATE KEY(id)
PARTITION BY LIST (dt,province) (
   PARTITION p1 VALUES IN (("2022-04-01", "beijing"),("2022-04-01", "chongqing")),
   PARTITION p2 VALUES IN (("2022-04-01", "shanghai"))
)
DISTRIBUTED BY HASH(id) BUCKETS 1
PROPERTIES(
    "replication_num" = "1"
);

3. Show create table for list partition

show create table t_recharge_detail_single;
show create table t_recharge_detail_multi;

liuyehcf committed:
Profile add metrics (#5457)

ZiheLiu committed:
Opt: set size of connection pool of jprotobuf to 16 by default (#5478)

The size of shared connection pool of jprotobuf is 1, which means protobuf RPC requests can only send sequentially. And the timeout for getting connection from the connection pool is 10s.
There are too many RPC requests for 1000-concurrency test. As a result, some requests timeout and add this BE to blacklist.

Modification

* Rename Config.brpc_number_of_concurrent_requests_processed to brpc_connection_pool_size, default is 16, to avoid * the existed users have already change brpc_number_of_concurrent_requests_processed in config file.
* Set the size of shared connection pool to Config.brpc_connection_pool_size.
* set MaxIdleSize to the size of shared connection pool for both shared and non-shared connection pool.
* After the RPC request sending, the connection will be closed, if the number of total connections is greater than MaxIdleSize. Therefore, MaxIdleSize shouldn't less than MaxTotal.
* Remove the branch enable_brpc_share_channel=false

Linkerist committed:
Just a tiny refactor&remove for repeated code, as a single PR incase of disturbing review (#5465)

zdsg1024 committed:
[SR GeekCamp] Support fe grammar for GBK charset (#5343)

There are following improvement in old parser:
1.status can support character_set_database variable
2.show collation can show gbk character
3.jdbc connection can support characterEncoding=gbk
4.create table can support charset for table level
5.create table can support charset for column level

Seaven committed:
Fix merge same group expression error (#5317)

xlwh committed:
[Geek Camp] Add tracing common util (#5287)

zddr committed:
[MaterializedView]Drop Materialized vied stmt (#4957)

mchades committed:
Add date_floor function (#5187)

itweixiang committed:
Encapsulate a tools for stream load when import file with a big size #4031 (#5451)

Use java and shell realize tools about stream load

The method of stream load is limited by `curl` performance. As a result, it's too slow to import a big file.
I want to realize a Java client to transfer the big file in parallel, can increase the performance.

First, I need to sequential read the big file, due to the characteristics of the disk, the performance of sequential read is of less concern about thread number.
Second, the data that was read will compose a long string, and be put in the blocking queue. when the main thread will block, the size of the blocking queue is equal to a limited size.
Third, the worker thread will consume the blocking queue, and send data through the network. when the worker thread is stopped, they receive an empty string from the main thread.

Using curl with one thread, the transmission speed is limited by disk bandwidth and at tens of megabytes.
With the new tools, the transmission speed can run over the 10 Gigabit network and satulated the disk bandwidth.
However, the transmission speed may also be affected by the sequential read-write performance of the hard disk. If you want faster transmission speed, you need a hard disk with better performance

sduzh committed:
Enable implicit cast to StringPiece from std::string_view (#5406)

satanson committed:
remove concurrency_limit check in resource group (#5507)

dirtysalt committed:
add -lgtest_main and -lbenchark_main for easily integrating ut (#5508)

sduzh committed:
Typofix: peak -> peek (#5496)

shyamrox committed:
Fixing Issues #5446 and 5435- Removing unused code. (#5510)

* Fixing Issue #5446 - Removing unused timer code.

* Fixing Issue# 5435 - TabletManager::get_partition_related_tablets can be deleted

liuyehcf committed:
Restore profile merge process changed by #4456 (#5503)

murphyatwork committed:
initialize column in create_column (#5170)

ColumnHelper::create_column has a parameter size, but only works for ConstColumn, which is error-prone.

This PR try to clarify this function:

* Parameter size is used to initialize the column with specified size and default value. This semantic is same as std::vector<T>(size_t size)
* The default value of each column is determined by the column itself. E.g. BinaryColumn's default value is empty string, NullableColumn is null.

murphyatwork committed:
avoid extend query deadline to implement a correct timeout (#5486)

Avoid extending deadline when executing query, which satisfy the query_timeout semantic.

ZiheLiu committed:
Reject coming fragment instances when BE overloaded (#5412)

stdpain committed:
use standard uuid format (#5492)

* use standard uuid format

```
mysql> select uuid();
+--------------------------------------+
| uuid()                               |
+--------------------------------------+
| 22e4147d-dd76-7df4-6ef3-001600000000 |
+--------------------------------------+
1 row in set (0.01 sec)
```

murphyatwork committed:
Reduce chunk_size in scan for huge length types (#5484)

Astralidea committed:
Fix bug priorNetwork inclusiveHostCount (#5525)

meegoo committed:
Optimize implementation of version graph (#5475)

1. Simplify the generation logic of the version graph
2/ Use the greedy algorithm to find the shortest version path, replacing the original BFS algorithm
In theory, there is only a containment relationship between the Versions of StarRocks, and there is no cross relationship.
   (0->0 1->1 2->2 0->2) exists
   (0->2 1->3) not exists
So that we can use the greedy algorithm other than BFS

decster committed:
Add log for slow database lock and fix some logging (#5453)

This PR adds warning log with some debug information when db's read/write lock takes a long time to lock, this helps identify lock contention issues. Also fix a logging error in TabletScheduler.

dirtysalt committed:
Add BE connector for DLA new framework (#5520)

Seaven committed:
Refactor optimizer join code (#5416)

DorianZheng committed:
[DLA New Framework] Refactor: rename Catalog to GlobalStateMgr (#5438)

Astralidea committed:
Fix some basic comment

Astralidea committed:
Fix some basic comment

Astralidea commented on fe/fe-core/src/main/java/com/starrocks/mv/MaterializedViewJobManager.java:
```diff
@@ -0,0 +1,340 @@
+                if (owner != null) {
+                    LOG.warn("materialized view lock is held by: {}", Util.dumpThread(owner, 50));
+                } else {
+                    LOG.warn("materialized view lock owner is null");
```
If a thread takes its own lock owner is null

Astralidea commented on fe/fe-core/src/main/java/com/starrocks/mv/MaterializedViewJobManager.java:
```diff
@@ -0,0 +1,340 @@
+        } catch (InterruptedException e) {
+            LOG.warn("got exception while getting materialized view lock", e);
+        }
+        return lock.isHeldByCurrentThread();
```
If a thread holds its own lock, it will be here

Astralidea commented on fe/fe-core/src/main/java/com/starrocks/mv/MaterializedViewJobExecutor.java:
```diff
@@ -0,0 +1,51 @@
+            job.generateTasks();
+            job.setStatus(Constants.MaterializedViewJobStatus.RUNNING);
+        } else {
+            job.setStatus(Constants.MaterializedViewJobStatus.RETRYING);
```
Because you have to let the user know if he is running it for the first time or if he is trying again.

Astralidea commented on fe/fe-core/src/main/java/com/starrocks/mv/MaterializedViewJobManager.java:
```diff
@@ -0,0 +1,340 @@
+                }
+
+                // put the pending job that can be run into running map
+                Iterator<Long> pendingIterator = pendingJobMap.keySet().iterator();
```
This has been concurrency control through the above tryLock

Astralidea commented on fe/fe-core/src/main/java/com/starrocks/mv/MaterializedViewJobManager.java:
```diff
@@ -0,0 +1,340 @@
+        long jobId = Catalog.getCurrentCatalog().getNextId();
+        job.setId(jobId);
+        long mvTableId = job.getMvTableId();
+        Queue<MaterializedViewRefreshJob> jobQueue = pendingJobMap.get(mvTableId);
```
This cannot use getOrDefault because if jobQueue is null it should be put a job in it. else it should merge the job.

Astralidea commented on fe/fe-core/src/main/java/com/starrocks/mv/MaterializedViewJobManager.java:
```diff
@@ -0,0 +1,340 @@
+    }
+
+    public void replayRegisterScheduledJob(MaterializedViewSchedulerInfo info) {
+        periodScheduledManager.put(info.getId(), info);
```
The replay operation is single-threaded and does not require concurrency control

zhuxt2015 committed:
Support alter table rename table clause in new Parser and new Analyzer (#5285)

Astralidea commented on fe/fe-core/src/main/java/com/starrocks/mv/MaterializedViewSchedulerInfo.java:
```diff
@@ -0,0 +1,78 @@
+public class MaterializedViewSchedulerInfo implements Writable {
+
+    @SerializedName("id")
+    private Long id;
```
because catalog.nextId returns long

Astralidea commented on fe/fe-core/src/main/java/com/starrocks/mv/MaterializedViewRefreshJobBuilder.java:
```diff
@@ -0,0 +1,89 @@
+        return this;
+    }
+
+    @VisibleForTesting
```
The meaning of this Tag is that this method should be private, and now it is public just for testing.

Astralidea commented on fe/fe-core/src/main/java/com/starrocks/mv/MaterializedViewRefreshJob.java:
```diff
@@ -0,0 +1,267 @@
+        tasks.add(task);
+    }
+
+    public void runTasks()  {
```
Because your refresh task is generated on demand when it is about to run. Instead of generating all tasks from the beginning.

Astralidea commented on fe/fe-core/src/main/java/com/starrocks/mv/MaterializedViewRefreshJob.java:
```diff
@@ -0,0 +1,267 @@
+            if (task.getStatus() == Constants.MaterializedViewTaskStatus.PENDING) {
+                task.beginTask();
+                task.setStatus(Constants.MaterializedViewTaskStatus.RUNNING);
+                try {
```
This exception needs to be captured at the task granularity. It is possible that one of your tasks failed, but other tasks can be executed successfully.

Astralidea commented on fe/fe-core/src/main/java/com/starrocks/mv/MaterializedViewSchedulerInfo.java:
```diff
@@ -0,0 +1,78 @@
+    private TimeUnit timeUnit;
+
+    @SerializedName("jobBuilder")
+    private MaterializedViewRefreshJobBuilder jobBuilder;
```
Because the user may submit a periodic Job, but the Job has not actually been created, this build is needed to create a specific Job.

wanpengfei-git commented:
#### [FE PR Coverage check]

:disappointed: **fail** : 314 / 486 (64.61%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/persist/EditLog.java|0|22|00.00%|
|:large_blue_circle:|com/starrocks/mv/MaterializedViewPartitionRefreshTask.java|0|19|00.00%|
|:large_blue_circle:|com/starrocks/journal/JournalEntity.java|0|9|00.00%|
|:large_blue_circle:|com/starrocks/mv/MaterializedViewRefreshJobStatusChange.java|6|23|26.09%|
|:large_blue_circle:|com/starrocks/mv/MaterializedViewRefreshTask.java|8|20|40.00%|
|:large_blue_circle:|com/starrocks/catalog/Catalog.java|1|2|50.00%|
|:large_blue_circle:|com/starrocks/mv/MaterializedViewSchedulerInfo.java|11|20|55.00%|
|:large_blue_circle:|com/starrocks/mv/MaterializedViewJobManager.java|139|198|70.20%|
|:large_blue_circle:|com/starrocks/mv/MaterializedViewJobExecutor.java|18|23|78.26%|
|:large_blue_circle:|com/starrocks/mv/MaterializedViewRefreshJob.java|77|94|81.91%|
|:large_blue_circle:|com/starrocks/mv/MaterializedViewTaskException.java|5|6|83.33%|
|:large_blue_circle:|com/starrocks/mv/MaterializedViewRefreshJobBuilder.java|29|30|96.67%|
|:large_blue_circle:|com/starrocks/statistic/Constants.java|20|20|100.00%|

sym-liuyang committed:
[SR GeekCamp] Support function array_cum_sum in StarRocks (#5471)

Astralidea committed:
Opt.

Astralidea commented on fe/fe-core/src/main/java/com/starrocks/mv/MaterializedViewJobExecutor.java:
```diff
@@ -0,0 +1,51 @@
+        Future<?> future = refreshTaskPool.submit(() -> {
+            try {
+                job.setStartTime(LocalDateTime.now());
+                job.runTasks();
```
Exception is including InterruptedException

Astralidea committed:
Opt.

Astralidea commented on fe/fe-core/src/main/java/com/starrocks/mv/MaterializedViewRefreshJob.java:
```diff
@@ -0,0 +1,267 @@
+        tasks.add(task);
+    }
+
+    public void runTasks()  {
```
Exception is including InterruptedException

Astralidea commented on fe/fe-core/src/main/java/com/starrocks/mv/MaterializedViewJobManager.java:
```diff
@@ -0,0 +1,340 @@
+    Map<Long, MaterializedViewRefreshJob> runningJobMap;
+    // first long is scheduledId, manage generate periodicity tasks, only support 1 schedule currently
+    Map<Long, MaterializedViewSchedulerInfo> periodScheduledManager;
+    Deque<MaterializedViewRefreshJob> jobHistory;
```
If encapsulated as an interface, there will be many methods in this interface.

Astralidea commented on fe/fe-core/src/main/java/com/starrocks/mv/MaterializedViewJobManager.java:
```diff
@@ -0,0 +1,340 @@
+
+    public boolean retryJob(Long jobId) {
+        // Generally the retried tasks are the most recent tasks
+        Iterator<MaterializedViewRefreshJob> jobIter = jobHistory.iterator();
```
this JobHistory is Deque. if use map we will manage sort by time?

Astralidea committed:
Opt.

Astralidea commented on fe/fe-core/src/main/java/com/starrocks/mv/MaterializedViewJobManager.java:
```diff
@@ -0,0 +1,340 @@
+
+        // for dispatch tasks
+        ScheduledExecutorService dispatchTaskScheduler = Executors.newScheduledThreadPool(1);
+        dispatchTaskScheduler.scheduleAtFixedRate(() -> {
```
you are right!

Astralidea committed:
rebase to main

Astralidea commented:
This close because commit issue, new discussion will in #5558
